tests/php/test-class-amp-theme-support.php

```diff
@@ -223,6 +233,33 @@ public function test_read_theme_support_and_support_args() {
 		$this->assertTrue( get_theme_support( AMP_Theme_Support::SLUG ) );
 	}
 
+	/**
+	 * Test theme_supports_reader_mode.
+	 *
+	 * @covers \AMP_Theme_Support::theme_supports_reader_mode.
+	 */
+	public function test_theme_supports_reader_mode() {
+		remove_all_filters( 'theme_root' );
+
+		add_theme_support( 'amp' );
+		AMP_Theme_Support::read_theme_support();
+
+		// The mode is Standard, and there is no /amp directory, so this should be false.
+		$this->assertFalse( AMP_Theme_Support::theme_supports_reader_mode() );
+
+		remove_theme_support( 'amp' );
+		AMP_Theme_Support::read_theme_support();
+
+		// The mode is Reader, but there is no /amp directory in the theme.
+		$this->assertFalse( AMP_Theme_Support::theme_supports_reader_mode() );
+
+		$this->amp_directory_in_theme = get_template_directory() . '/amp';
+		mkdir( $this->amp_directory_in_theme );
```

Comment:
This makes me wary. It would be preferable for filesystem changes made during tests to be done in the temp directory. 

Could there perhaps instead be new mock themes created  in `tests/data/themes` similar to how [there is in core](https://github.com/WordPress/wordpress-develop/tree/master/tests/phpunit/data/themedir1)? There could be a theme in standard mode, a theme in transitional mode, and a theme in reader mode.

Activating such a theme could be done with something like this:

```php
register_theme_directory( __DIR___ . '/data/themes' );
search_theme_directories( true ); // Regenerate the transient.
switch_theme( 'theme-with-reader-mode' );
```

Having such test data themes in the plugin will be helpful for the future to reduce current dependencies we have on core themes being installed in the test suite.

@schlessera Thoughts?

Reply:
Yes, that could work and would be preferable to writing directly into a theme here.

Ideally, though, we wouldn't need the filesystem at all for unit tests. A cleaner approach would be to use a virtual filesystem that we fully control and can reset whenever we want using vfsStream: http://vfs.bovigo.org/

This has the following added benefits:
- Faster, everything happens in memory.
- No thrashing of the disk with temp files, so better for the disk health.
- Perfectly safe, as no operations hit the actual disk, so you don't risk a rogue delete operation to cause harm.
- No permissions issues, as you control the entire disk including the permissions, so more reliable across environments.
- Possibility to set up multiple differing filesystems, and switch as needed.

Reply:
Actually, as we don't have unit tests here, I'm not sure if we can make this work transparently with WordPress Core, so my above comment might not apply. We'd have to try if you can actually register a custom theme folder within that virtual file system without hitting issues with WordPress.

Reply:
Thanks, vfsStream actually seemed to work really well, though maybe my implementation could use some changes. How does https://github.com/ampproject/amp-wp/pull/3188/commits/634e5b09535c1deec5599570614ffbee6ce2c022 look?